File: src/quodeq/core/types/scan.py
"""Scan metadata types for the quick-scan phase."""

from __future__ import annotations

from dataclasses import dataclass, field


@dataclass(frozen=True, slots=True)
class ScanData:
    """Result of a quick project scan (no AI evaluation)."""

    file_tree: list[str] = field(default_factory=list)
    languages: dict[str, int] = field(default_factory=dict)
    branches: list[str] = field(default_factory=list)
    modules: list[str] = field(default_factory=list)
    scanned_at: str = ""
    total_files: int = 0
    # Subset of total_files whose extension is recognised as analysable
    # source code (matches the language ext-map used by the evaluation).
    code_files: int = 0

File: src/quodeq/ui/src/models/scan.js
/**
 * Scan-result model — a project's quick-scan summary (file/code/language/
 * branch counts). The raw API response (`ScanData` in `core/types/scan.py`,
 * served via `dataclasses.asdict` by `/scan` and `/projects/:id/scan`)
 * spells its fields snake_case.
 *
 * @typedef {Object} ScanSummary
 * @property {number}                totalFiles
 * @property {number}                codeFiles
 * @property {number}                untrackedFiles
 * @property {Object<string,number>} languages
 * @property {string[]}              branches
 * @property {string[]}              modules
 * @property {string}                scannedAt
 */

import { fromFieldSpec } from './fieldSpec.js';

/**
 * Field table for {@link createScanSummary}: output key -> [raw spelling(s), default].
 */
const SCAN_SUMMARY_FIELDS = {
  totalFiles:     [['totalFiles', 'total_files'], 0],
  codeFiles:      [['codeFiles', 'code_files'], 0],
  untrackedFiles: [['untrackedFiles', 'untracked_files'], 0],
  languages:      ['languages', () => ({})],
  branches:       ['branches', () => []],
  modules:        ['modules', () => []],
  scannedAt:      [['scannedAt', 'scanned_at'], ''],
};

/**
 * Create a canonical ScanSummary from a raw scan API response.
 *
 * @param {Object} raw
 * @returns {ScanSummary}
 */
export function createScanSummary(raw) {
  if (!raw || typeof raw !== 'object') return raw;
  return fromFieldSpec(raw, SCAN_SUMMARY_FIELDS);
}
